### ISSUE TITLE
ESP32PWM::allocatenext() Enters an Infinite Loop During Servo1.attach() on ESP32-S3

### ISSUE THREAD

**Issue by darkgrue** [madhephaestus/ESP32Servo#101]
VSCode PlatformIO:
-   platform = https://github.com/pioarduino/platform-espressif32.git#55.03.37 ; pioarduino, https://github.com/pioarduino/platform-espressif32
-   framework = arduino
-   board = esp32-s3-devkitm-1 ; EarGear 2 v4.1 Platform (ESP32-S3-MINI-1)
-   madhephaestus/ESP32Servo @ 3.1.3 ; https://registry.platformio.org/libraries/madhephaestus/ESP32Servo
-   arminjo/ServoEasing @ 3.6.0 ; https://registry.platformio.org/libraries/arminjo/ServoEasing

On ESP-IDF v5, the legacy `mcpwm_init()` internally tries to acquire a semaphore/mutex to serialize peripheral initialization. If this is called before the FreeRTOS scheduler has started, the blocking call inside `mcpwm_init()` waits indefinitely for a semaphore that the idle task would normally release - causing a freeze that looks like a hang at `Servo1.attach()`.

Calling `ESP32PWM::allocateTimer()` before `Servo1.attach()` _should_ put the library into explicit allocation mode, which disables the MCPWM preference path on ESP32-S3 and force LEDC-only allocation:

```
void initServos()
{
  uint8_t servo1Channel, servo2Channel;

  // Force LEDC-only timer allocation on ESP32-S3 to avoid the
  // legacy mcpwm_init() blocking call in ESP-IDF v5 / Arduino-ESP32 v3.x.
  ESP32PWM::allocateTimer(0);

  servo1Channel = Servo1.attach(PIN_SERVO1);
  servo2Channel = Servo2.attach(PIN_SERVO2);
  ...
```
However, in `allocatenext()`, the non-variable-frequency path (`useVariableFrequency = false`) tries MCPWM before LEDC regardless of explicit-allocation mode on ESP32-S3 (lines 157–185). So calling `allocateTimer()` has no effect on the MCPWM preference guard.

The quick fix is in `ESP32PWM.cpp:158`, add a guard so that when explicit allocation mode is active (i.e., `allocateTimer()` was called), the MCPWM preference is skipped and LEDC is used directly:

```
			// non-variable, prefer MCPWM with shared freq, fallback to LEDC
#if defined(CONFIG_IDF_TARGET_ESP32S3)
      if (!explicateAllocationMode) { // Skip MCPWM when explicit mode is set
			for(int u=0; u<MCPWM_NUM_UNITS; u++) {
				for(int t=0; t<MCPWM_NUM_TIMERS_PER_UNIT; t++) {
					bool freqMatch = (mcpwmTimers[u][t].freq == freqlocal || mcpwmTimers[u][t].freq == -1);
					if (freqMatch && mcpwmTimers[u][t].operatorCount < MCPWM_NUM_OPERATORS_PER_TIMER) {
						if (mcpwmTimers[u][t].freq == -1) {
							mcpwmTimers[u][t].freq = freqlocal;
						}
						mcpwmUnit = (mcpwm_unit_t)u;
						mcpwmTimer = (mcpwm_timer_t)t;
						int opIndex = mcpwmTimers[u][t].operatorCount;
						mcpwmOperator = (opIndex == 0) ? MCPWM_OPR_A : MCPWM_OPR_B;
						mcpwmTimers[u][t].operators[opIndex] = this;
						mcpwmTimers[u][t].operatorCount++;
						PWMCount++;
						myFreq = freq;
						isMCPWM = true;
						pwmChannel = -1;
						timerNum = -1;
						return 0;
					}
				}
			}
      } // Close the guard
#endif
			// if no MCPWM available, fallback to LEDC shared freq
```

Also `explicateAllocationMode` is misspelt. It should be `explicitAllocationMode`, unless I'm misunderstanding the intent? But as it's spelled that way elsewhere so I left it as-is to keep changes minimal.

A better fix would be to migrate to the new MCPWM driver (driver/mcpwm_prelude.h, available since ESP-IDF v5.0) The new driver's initialization calls do not wait on FreeRTOS synchronization primitives at the point of peripheral setup - they use spinlocks (`portENTER_CRITICAL`) for thread safety at runtime, which work correctly before the scheduler starts. But that would require all the library MCPWM code paths needing to be rewritten - the changes for the new driver are substantial.

**Comment by jks-liu** [comment]
same error

**Comment by nGoline** [comment]
Mine is a little bit different.

VSCode PlatformIO:
- platform = https://github.com/pioarduino/platform-espressif32.git#55.03.37
- framework = arduino
- board = esp32s3n16r8 ;
- madhephaestus/ESP32Servo @ 3.1.3 ; https://registry.platformio.org/libraries/madhephaestus/ESP32Servo
- adafruit/Adafruit NeoPixel @ ^1.15.3

```cpp
void configureServo()
{
  ESP32PWM::allocateTimer(0);
  ESP32PWM::allocateTimer(1);
  ESP32PWM::allocateTimer(2);
  ESP32PWM::allocateTimer(3);
  attachServo();
  myservo.write(previousWaitingPosition);
  yieldDelay(500);
  detachServo();
}

void attachServo()
{
  if (!myservo.attached())
  {
    myservo.attach(SERVO_PIN, 500, 2700);
  }
}

void detachServo()
{
  if (myservo.attached())
  {
    myservo.detach();
  }
}
```

I can see on the logs that the library tries to attach the servo `[ESP32Servo.cpp:84] attach(): [ESP32Servo] Attempting to Attach servo on pin=17 min=500 max=270` and enters a infinite loop.

I added logs to several parts and found out that on `int Servo::attach(int pin, int min, int max):133` it calls `pwm.attachPin(this->pinNumber,REFRESH_CPS, this->timer_width );`, then `int ret=setup(freq, resolution_bits);` and, since my `isMCPWM` variable is false, finally `checkFrequencyForSideEffects(freq);`. My `getTimer()` returns `-1`, which makes my `for (int i = 0; i < timerCount[getTimer()]; i++) {` loop infinite.

I've kind of fixed it by adding this check:
```
allocatenext(freq);
// \/ Added the check here \/
if(getTimer() < 0)
    return true;
```

Now my servo works even though an error is printed out.

Unfortunately I don't have time to debug how it gets into this state, but I plan to get back to this in the future if no one finds a real solution.
